File: lambdas/ride/ride.py
import os
import boto3
from boto3.dynamodb.conditions import Key
import json

RECENT_TRIPS_TABLE = os.getenv('RECENT_TRIPS_TABLE', 'TripsTable')
PAGE_LIMIT = int(os.getenv('PAGE_LIMIT', 5))  # Ensure PAGE_LIMIT is an integer

dynamodb = boto3.resource('dynamodb')

def handler(event, context):
    print(f'Received event: {event}')
    
    headers = {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*',  # Allow requests from any origin
        'Access-Control-Allow-Methods': 'OPTIONS,GET',  # Allow GET and OPTIONS methods
        'Access-Control-Allow-Headers': 'Content-Type,Authorization',  # Allow specific headers
        'Access-Control-Allow-Credentials': 'true'
    }
    
    # Handle CORS preflight requests
    if event['httpMethod'] == 'OPTIONS':
        return {
            'statusCode': 200,
            'headers': headers,
            'body': json.dumps({'message': 'CORS preflight successful'})
        }

    # Handle GET requests
    if event['httpMethod'] == 'GET':
        try:
            user_id = event['queryStringParameters'].get('userId')
            if not user_id:
                return {
                    'statusCode': 400,
                    'headers': headers,
                    'body': json.dumps({'message': 'userId parameter is required'})
                }

            limit = int(event['queryStringParameters'].get('limit', PAGE_LIMIT))
            last_evaluated_key = event['queryStringParameters'].get('lastEvaluatedKey')
    
            table = dynamodb.Table(RECENT_TRIPS_TABLE)
    
            # Building the query parameters
            query_params = {
                'KeyConditionExpression': Key('userId').eq(user_id),
                'Limit': limit
            }
    
            # Add the ExclusiveStartKey if pagination is being used
            if last_evaluated_key:
                query_params['ExclusiveStartKey'] = json.loads(last_evaluated_key)
    
            # Query the DynamoDB table
            response = table.query(**query_params)
    
            # Prepare the result with pagination information
            result = {
                'trips': response['Items']
            }
    
            # If there are more items to be fetched, include the LastEvaluatedKey
            if 'LastEvaluatedKey' in response:
                result['lastEvaluatedKey'] = json.dumps(response['LastEvaluatedKey'])
    
            return {
                'statusCode': 200,
                'headers': headers,
                'body': json.dumps(result)
            }
        except Exception as e:
            print(f"Exception occurred: {str(e)}")
            return {
                'statusCode': 500,
                'headers': headers,
                'body': json.dumps({'message': 'Error fetching trips', 'error': str(e)})
            }
    else:
        return {
            'statusCode': 405,
            'headers': headers,
            'body': json.dumps({'message': 'Method not allowed'})
        }

File: lambdas/src/ride/ride.ts
import { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda';
import * as AWS from 'aws-sdk';

const RECENT_TRIPS_TABLE = process.env.RECENT_TRIPS_TABLE || 'TripsTable';
const PAGE_LIMIT = Number(process.env.PAGE_LIMIT) || 5;

const dynamodb = new AWS.DynamoDB.DocumentClient();

export const handler = async (event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> => {
    const headers = {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Methods': 'OPTIONS,GET',
        'Access-Control-Allow-Headers': 'Content-Type,Authorization',
        'Access-Control-Allow-Credentials': 'true',
    };

    // Handle CORS preflight requests
    if (event.httpMethod === 'OPTIONS') {
        return {
            statusCode: 200,
            headers,
            body: JSON.stringify({ message: 'CORS preflight successful' }),
        };
    }

    // Handle GET requests
    if (event.httpMethod === 'GET') {
        try {
            const userId = event.queryStringParameters?.userId;
            if (!userId) {
                return {
                    statusCode: 400,
                    headers,
                    body: JSON.stringify({ message: 'userId parameter is required' }),
                };
            }

            const limit = Number(event.queryStringParameters?.limit) || PAGE_LIMIT;
            const lastEvaluatedKey = event.queryStringParameters?.lastEvaluatedKey ? JSON.parse(event.queryStringParameters.lastEvaluatedKey) : undefined;

            const queryParams: AWS.DynamoDB.DocumentClient.QueryInput = {
                TableName: RECENT_TRIPS_TABLE,
                KeyConditionExpression: 'userId = :userId',
                ExpressionAttributeValues: {
                    ':userId': userId,
                },
                Limit: limit,
                ScanIndexForward: false,  // Get latest trips first
            };

            // Add the ExclusiveStartKey if pagination is being used
            if (lastEvaluatedKey) {
                queryParams.ExclusiveStartKey = lastEvaluatedKey;
            }

            const response = await dynamodb.query(queryParams).promise();

            const result = {
                trips: response.Items?.map(item => {
                    return {
                        ...item,
                        // You may handle any additional decimal to float conversion here if needed
                    };
                }) || [],
            } as any;

            // If there are more items to be fetched, include the LastEvaluatedKey
            if (response.LastEvaluatedKey) {
                result['lastEvaluatedKey'] = JSON.stringify(response.LastEvaluatedKey);
            }

            return {
                statusCode: 200,
                headers,
                body: JSON.stringify(result),
            };
        } catch (error) {
            console.error(`Exception occurred: ${error}`);
            return {
                statusCode: 500,
                headers,
                body: JSON.stringify({ message: 'Error fetching trips', error: (error as Error).message }),
            };
        }
    } else {
        return {
            statusCode: 405,
            headers,
            body: JSON.stringify({ message: 'Method not allowed' }),
        };
    }
};
